src/wallet/coinselection.cpp

```diff
@@ -335,23 +338,29 @@ util::Result<SelectionResult> KnapsackSolver(std::vector<OutputGroup>& groups, c
         for (const auto& group : applicable_groups) {
             result.AddInput(group);
         }
-        return result;
+        if (result.GetWeight() <= max_input_weight) return result;
+
+        // Try something else
+        result.Clear();
     }
 
     if (nTotalLower < nTargetValue) {
-        if (!lowest_larger) return util::Error();
+        if (!lowest_larger) return util::Error{Untranslated(strprintf("Only %d coins to cover target value of %d", nTotalLower, nTargetValue))};
         result.AddInput(*lowest_larger);
-        return result;
+        if (result.GetWeight() <= max_input_weight) return result;
```

Comment:
nit: `lowest_larger` can't be above `max_input_weight`. You already checked it above.

I prefer if we filter above weight inputs once at the top of the function.